### PR TITLE
Enable strict mode for DEBUG builds to detect unclosed cursor/stream

### LINKED ISSUE #2924: Add a code quality check to prevent leaking cursors
#### Problem description
I noticed that at some of the places, the cursors weren't being closed after usage. Sometimes, the diff is so huge that it becomes really hard to catch this leak. Having a automatic check would prevent this from happening again in future. Related PRs https://github.com/opendatakit/collect/pull/2911 https://github.com/opendatakit/collect/pull/2912

#### Expected behavior
CircleCI should fail the build if any unclosed cursor is found during static code analysis

#### Other information 
I guess there are already some rules present in the static code analysis toolset which can be utilized. 

@opendatakit-bot claim

### PR DESCRIPTION
**Pull request by shobhitagarwal1612**
Closes #2924 

#### What has been done to verify that this works as intended?
At commit https://github.com/opendatakit/collect/commit/a4b04bbcf216c2d2db03a4e58581be1950396c55, the app crashes while trying to check if the form is already available, downloading blank form, importing settings using image saved in gallery or loading form from the cache. This is because, at those places, the cursor/stream isn't being closed after usage.

Tests performed to ensure that the new changes don't cause any regressions, 
 - Downloading blank form
 - Refresh download list multiple times
 - Loading blank form multiple times
 - Importing settings from SD Card

#### Why is this the best possible solution? Were any other approaches considered?
Apart from adding the [strict mode](https://developer.android.com/reference/android/os/StrictMode.html), I have also updated the PMD plugin to 6.12.0 as that includes the check to detect usage of `try-with-resources`. That can be used to later convert usages to further prevent the introduction of such regressions due to a missed close statement.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
Might affect the form/media downloading, form loading from cache or importing settings. Must be verified thoroughly before merging

#### Do we need any specific form for testing your changes? If so, please attach one.
Any form can be used. I used the `All-Widgets-Form`

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/opendatakit/docs/issues/new) and include the link below.
No, whenever the crash happens, an appropriate log is also displayed in the logcat with all necessary details.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

### CODE REVIEW

**Comment by shobhitagarwal1612** [comment]
The tests are failing due to the introduction of the strict mode in DEBUG builds. Should we run `release` unit and instrumentation tests instead of `debug` from now on? 
**CC** @grzesiek2010

**Comment by shobhitagarwal1612** [comment]
@grzesiek2010 I have attempted to enable the strict mode in debug only mode, but that causes robolectric tests to fail. I have been tried different ways to do it but nothing has worked so far. Hence, I propose that we should scope this PR as follows:
 - Fix for the caught unclosed cursor/streams
 - PMD update 
 - Commented strict mode with instruction in README to uncomment it while testing on a real device
 - Undo everything else

@cooperka @grzesiek2010 @zestyping Please comment your thoughts

**Comment by grzesiek2010** [comment]
That's strange it doesn't work with Robolectric...
What about not enabling strict mode in that case? You could use: 
```java
if (!"robolectric".equals(Build.FINGERPRINT)) {
            StrictMode.setThreadPolicy(new StrictMode.ThreadPolicy.Builder()
                    .detectAll()
                    .penaltyLog()
                    .penaltyDeath()
                    .build());

            StrictMode.setVmPolicy(new StrictMode.VmPolicy.Builder()
                    .detectLeakedSqlLiteObjects()
                    .detectLeakedClosableObjects()
                    .penaltyLog()
                    .penaltyDeath()
                    .build());
        }
```

**Comment by shobhitagarwal1612** [comment]
> Build.FINGERPRINT

This returns the fingerprint of the Android OS and not the app's build. 
This is what it shows for my device Redmi Note 4:
`xiaomi/mido/mido:7.0/NRD90M/V10.2.3.0.NCFMIXM:user/release-keys`

**Comment by shobhitagarwal1612** [comment]
@grzesiek2010 
The problem why my initial approach didn't work was because I used `debuggable` instead of the type of build when checking for the type of the current build. So, it was always enabled when running unit/instrumentation tests.

Please review now

**Comment by grzesiek2010** [comment]
> This returns the fingerprint of the Android OS and not the app's build.
This is what it shows for my device Redmi Note 4:
xiaomi/mido/mido:7.0/NRD90M/V10.2.3.0.NCFMIXM:user/release-keys

yes, but if you run a robolectric test it returns "robolectric" so you can avoid enabling it. I think it's the easiest solution instead of running release tests - it would require discussion.

**Comment by shobhitagarwal1612** [comment]
@grzesiek2010 I have fixed the commit history and applied your solution. Please review now.

File: collect_app/src/main/java/org/odk/collect/android/application/Collect.java
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/application/Collect.java]: Why `!BuildConfig.BUILD_TYPE.equals("debug")` ?

Comment (shobhitagarwal1612) [reply collect_app/src/main/java/org/odk/collect/android/application/Collect.java]: So that if someone forks the repo and wants to see the debug logs, then it works but doesn't enable strict mode as well which might not be useful to others

Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/application/Collect.java]: We have three build variants:
odkCollectRelease - our releases
release - fork releases 
debug

so, in this case, it's enabled in `release` but should be in `debug`. Am I missing something?

Comment (shobhitagarwal1612) [reply collect_app/src/main/java/org/odk/collect/android/application/Collect.java]: It is only enabled in `debug`. The `robolectric` build check is nested within DEBUG block.

odkCollectRelease - strict mode disabled
release - strict mode disabled
debug - strict mode enabled

Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/application/Collect.java]: But in your if statement you have `not equals debug`

Comment (shobhitagarwal1612) [reply collect_app/src/main/java/org/odk/collect/android/application/Collect.java]: Oops, my bad. :grin:

Comment (shobhitagarwal1612) [reply collect_app/src/main/java/org/odk/collect/android/application/Collect.java]: Fixed

File: collect_app/src/main/java/org/odk/collect/android/http/HttpClientConnection.java
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/http/HttpClientConnection.java]: Why do you use the new constructor only here but not in other classes which implements `OpenRosaHttpInterface`?

Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/http/HttpClientConnection.java]: We use Httpclient in other methods in this class as well but you edited only this one, why?

Comment (shobhitagarwal1612) [reply collect_app/src/main/java/org/odk/collect/android/http/HttpClientConnection.java]: These warnings were resolved based on the detections reported by strict mode.

Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/http/HttpClientConnection.java]: Maybe you just used that one method testing the app manually. It would be strange that in one place `httpclient` should be closed but in another place not. Could you investigate it deeper?

Comment (shobhitagarwal1612) [reply collect_app/src/main/java/org/odk/collect/android/http/HttpClientConnection.java]: I thought of the same, but then I noticed that this will soon be replaced with OkHttp implementation. So, we should not spend too much time improving obsolete codebase. Moreover, if cursor is actually not getting closed, then we will get notified for the debug builds.

Do you think it's worth investigating deeper into this given the above statements?

Comment (shobhitagarwal1612) [reply collect_app/src/main/java/org/odk/collect/android/http/HttpClientConnection.java]: Let's close the cursor elsewhere as well. Probably won't take much time. :+1:

Comment (shobhitagarwal1612) [reply collect_app/src/main/java/org/odk/collect/android/http/HttpClientConnection.java]: Just noticed that OkHttp implemententation also uses the same Closeable response. So it is gonna happen in the new code as well

Comment (shobhitagarwal1612) [reply collect_app/src/main/java/org/odk/collect/android/http/HttpClientConnection.java]: On deeper investigation I found out that `CloseableHttpClient` is auto-closed at the end of `execute` method. There is some issue with the custom http client implementation(httpclientandroidlib-4.5.2-2.jar). The auto-close doesn't work. Even calling the `close()` method explicitly doesn't work at all. On the contrary, with `OkHttpConnection` class, explicitly calling the `close()` method works as expected

So, let's merge this PR only only after #2869 is resolved otherwise the debug app will continuously crash due to the strict mode

Comment (shobhitagarwal1612) [reply collect_app/src/main/java/org/odk/collect/android/http/HttpClientConnection.java]: To validate, try fetching form list with the current PR 2-3 times. You'll notice that second time, the crash will occur.

Then replace `HttpClientConnection` with `OkHttpConnection` in `AppDependencyModule` and perform the same tasks again. Now you won't notice any such crash

Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/http/HttpClientConnection.java]: @lognaturel when will we fix #2869?

Comment (shobhitagarwal1612) [reply collect_app/src/main/java/org/odk/collect/android/http/HttpClientConnection.java]: We can remove/comment the `.penaltyDeath()` to prevent app crash and only display the error on the log if that PR is going to be merged in future

Comment (lognaturel) [reply collect_app/src/main/java/org/odk/collect/android/http/HttpClientConnection.java]: Afraid there's no timeline for #2869 at the moment.

Comment (shobhitagarwal1612) [reply collect_app/src/main/java/org/odk/collect/android/http/HttpClientConnection.java]: What's blocking it?

File: collect_app/src/main/java/org/odk/collect/android/http/OkHttpConnection.java
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/http/OkHttpConnection.java]: It feels like an odd design to me that the response is passed to `HttpGetResult` only so it can be closed. The response isn't actually used in `HttpGetResult` so in both cases, it looks like  the response could be closed before the result is returned. Is there a reason you did it this way?

**Review by lognaturel — COMMENTED** [review]
Thanks for updating this one, @shobhitagarwal1612. It seems like a good thing to do to me. Just one question inline.

File: collect_app/src/main/java/org/odk/collect/android/http/HttpClientConnection.java
Comment (lognaturel) [reply collect_app/src/main/java/org/odk/collect/android/http/HttpClientConnection.java]: To answer your old question, it was blocked on getting some tests in place and fixing any remaining small issues which @seadowg did!

**Review by seadowg — CHANGES_REQUESTED** [review]
@shobhitagarwal1612 is this something you're going to be able to get back to soon or should we close for now?

**Comment by shobhitagarwal1612** [comment]
Sorry, I completely forgot about this PR. I'll give it another go during the weekend if that's fine with everyone

File: collect_app/src/main/java/org/odk/collect/android/http/openrosa/HttpGetResult.java
Comment (seadowg) [inline collect_app/src/main/java/org/odk/collect/android/http/openrosa/HttpGetResult.java]: Why do we pass in the `Closeable` here?

File: collect_app/src/main/java/org/odk/collect/android/application/Collect.java
Comment (seadowg) [inline collect_app/src/main/java/org/odk/collect/android/application/Collect.java]: Huh I didn't know you could use `FINGERPRINT` for this. I think I'd actually prefer we use the `TestCollect` class that Robolectric automatically references for customizing the test application. This keeps test code out of our `Collect` class.

**Review by seadowg — CHANGES_REQUESTED** [review]

**Comment by seadowg** [comment]
I'm just going to close this for now as it's been inactive for a while but obviously feel free to reopen if you get started on it again @shobhitagarwal1612! Miss you 👋

**Comment by shobhitagarwal1612** [comment]
@seadowg A lot had happened since my last update on this branch. So, I basically rebased it onto upstream/master and cleaned it up a bit. PTAL!

The strict mode only logs to logcat instead of crashing the entire app. So, the remaining ones can be dealt with separately

File: collect_app/src/main/java/org/odk/collect/android/application/Collect.java
Comment (seadowg) [inline collect_app/src/main/java/org/odk/collect/android/application/Collect.java]: I'm a little worried that this will get lost as Logcat is really noisy. Could we maybe use `penaltyDialog`? Given it's only on for debug builds I think it's fine for it to be loud!

**Review by seadowg — DISMISSED** [review]

**Review by seadowg — CHANGES_REQUESTED** [review]
As comment

File: collect_app/src/main/java/org/odk/collect/android/application/Collect.java
Comment (shobhitagarwal1612) [reply collect_app/src/main/java/org/odk/collect/android/application/Collect.java]: I tried that, but there were too many dialogs getting generated as soon as the app opened and caused the app to crash. Maybe we should try that once the known leaks are fixed. Thoughts?

Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/application/Collect.java]: Could we fix those problems as part of this PR? I think if we're going to put a check in we should fix any existing problems the check would find.

Comment (shobhitagarwal1612) [reply collect_app/src/main/java/org/odk/collect/android/application/Collect.java]: 👍

Comment (lognaturel) [reply collect_app/src/main/java/org/odk/collect/android/application/Collect.java]: Thanks for picking this back up, @shobhitagarwal1612!

Agreed that would be ideal. That said, if it requires really invasive refactoring it may not be appropriate. Could you start by enumerating the issues you've identified and maybe a quick estimate of whether you think it's a straightforward fix?

Comment (lognaturel) [reply collect_app/src/main/java/org/odk/collect/android/application/Collect.java]: I also suspect that https://github.com/opendatakit/collect/issues/3705 would fix most of the issues occurring on load. Please see what you think, @shobhitagarwal1612, so we can avoid duplicate/conflicting changes.

Comment (shobhitagarwal1612) [reply collect_app/src/main/java/org/odk/collect/android/application/Collect.java]: Looks like there's a leak in `osmdroid` library itself which we can't do anything about. 

Secondly, a lot disk read operations (shared prefs) happen on UI thread. That can be solved by moving all those calls to a non-UI thread. This looks like a big refactor to me

Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/application/Collect.java]: Hmmm interesting that it picks up the shared prefs stuff. I'd say as long as we're calling `apply` instead of `commit` I'm not really worried about it for now (did some reason [here](https://stackoverflow.com/questions/4371273/should-accessing-sharedpreferences-be-done-off-the-ui-thread)). I think we can probably disable that check if possible?

**Comment by lognaturel** [comment]
I'd like to merge this as-is. The leak fixes are helpful. I agree that the notices will both get lost in the existing log noise and add to that noise but I think that it's common to put a filter when looking at logcat output (right?). This means that if we have a PR with lots of new strings we can easily add a "leak" filter to logcat while trying it out. I'm not actually seeing the shared pref logs for some reason but even if there were noise added, I think a reminder to consider leaks is not bad. Like @shobhitagarwal1612 says, using dialogs is off the table because of the library leaks.

@seadowg?

**Review by lognaturel — APPROVED** [review]

**Review by seadowg — APPROVED** [review]
I agree with @lognaturel here. I think we can maybe iterate on this in future to make it crash in dev/test (once we've fixed any exisiting issues). For another time though!

@shobhitagarwal1612 looks like there is a conflict to fix but then we're all good!

@getodk/testers for this one I think it would be good to check opening and reopening forms (as there are changes to the form def cache mechanism) and downloading forms from a server.

File: collect_app/src/main/java/org/odk/collect/android/application/Collect.java
Comment (shobhitagarwal1612) [reply collect_app/src/main/java/org/odk/collect/android/application/Collect.java]: >  I think we can probably disable that check if possible?

Done

**Comment by lognaturel** [comment]
Thanks so much, @shobhitagarwal1612, for getting right on that! 🤩

**Comment by lognaturel** [comment]
Merging without QA because the diff is so easy to review and it's low risk anyway. I think the three checks needed will happen in the regression checks anyway but would be good to confirm, @getodk/testers:
- downloading a form, updating it on the server, seeing an update
- opening a form 3 times
- reading a configuration QR code

**Comment by mmarciniak90** [comment]
**Tested with success**

Verified on Android: 6.0, 8.1, 9.0
Verified mentioned cases.
Verified on the master branch.

@getodk-bot unlabel "needs testing"
@getodk-bot label "behavior verified"

**Pull request merged**
